### ISSUE TITLE
Subscriptions

### ISSUE THREAD

**Issue by bgentry** [bgentry/ember-apollo-client#4]
This addon should have built-in support for [GraphQL Subscriptions](https://dev-blog.apollodata.com/graphql-subscriptions-in-apollo-client-9a2457f015fb#.95gf6q85r) in order to support real-time streaming updates of records.

See also: http://graphql.org/blog/subscriptions-in-graphql-and-relay/

**Comment by joshuataylor** [comment]
This library is great, really really great.

Looking at different servers, it looks like each does things slightly differently for subscriptions, in terms of transport.

For my backend, I'm using Elixir, with Absinthe, and they use websockets. They have [a library for apollo](https://github.com/vic/apollo-phoenix-websocket), and this looks like I would just override the service to support it in my app, or we could create another app that will bridge that and this repo.

What are you thoughts on subscriptions, are you using them? Would love to help out if I have enough knowledge for this.

**Comment by xtagon** [comment]
Thanks for sharing that, @joshuataylor. I am also using Elixir and Absinthe and would love to figure out how to bridge that with Ember, especially for subscriptions! Maybe this deserves its own issue.

**Comment by viniciussbs** [comment]
It looks like you need to create a different interface for each subscription using this library. If this is the path, you'll need to implement your own service extending Apollo service just for subscriptions and change the interface on the fly - to avoid having one client per interface.

I don't know if what I'm talking makes sense.

**Comment by viniciussbs** [comment]
And maybe you will need to webpack this library in order to import the files, I guess.

**Comment by bgentry** [comment]
> This library is great, really really great.

Thanks @joshuataylor! Hoping to take it to the next level pretty soon :)

> Looking at different servers, it looks like each does things slightly differently for subscriptions, in terms of transport.

I haven't looked at this in a couple months, but this was my take as of then. Any subscription support in this library would need to be general enough to work with any of these backends (at least the common ones).

> What are you thoughts on subscriptions, are you using them? Would love to help out if I have enough knowledge for this.

I'm not yet, but only because it hasn't been important enough for me yet to devote the time to figuring it out. I definitely want them!

Maybe it's just a matter of looking at what the [React integration](http://dev.apollodata.com/react/subscriptions.html) does, specifically using [`subscriptions-transport-ws`](https://github.com/apollographql/subscriptions-transport-ws) with a custom network interface to add subscription support.

Definitely would love for anybody who has the time/interest to help pave the way here 🙏

**Comment by ArnaudValensi** [comment]
If someone need it, here is how I do subscription with ember-apollo-client:

```js
// apollo-service.js
import Ember from 'ember';
import ApolloService from 'ember-apollo-client/services/apollo';
import subscriptionsTransportWs from 'npm:subscriptions-transport-ws';
import { createNetworkInterface } from 'apollo-client';
import copyWithExtras from 'ember-apollo-client/utils/copy-with-extras';

const { SubscriptionClient, addGraphQLSubscriptions } = subscriptionsTransportWs;
const {
  computed,
  merge,
  Object: EmberObject,
  run,
  setProperties
} = Ember;

export default ApolloService.extend({
  session: Ember.inject.service('session'),

  clientOptions: computed(function () {
    let opts = this._super(...arguments);

    // Create regular NetworkInterface by using apollo-client's API:
    const networkInterface = createNetworkInterface({
      uri: 'http://localhost:3001/graphql'
    });

    // Create WebSocket client
    const wsClient = new SubscriptionClient(`ws://localhost:3001/subscriptions`, {
      reconnect: true
    });

    const networkInterfaceWithSubscriptions = addGraphQLSubscriptions(
      networkInterface,
      wsClient
    );

    return merge(opts, {
      networkInterface: networkInterfaceWithSubscriptions
    });
  }),

  subscribe(query, variables = {}, propertyName = null) {
    const apolloClient = this.get('client');

    let observable = apolloClient.subscribe({ query, variables });
    let subscription;

    let obj = EmberObject.create({
      unsubscribe() {
        subscription.unsubscribe();
      },
      [propertyName]: EmberObject.create({})
    });

    subscription = observable.subscribe({
      next: (rawData) => {
        if (rawData === undefined) {
          return;
        }

        let data = rawData[Object.keys(rawData)[0]];
        let dataToSend = copyWithExtras(data, [], []);

        run(() => setProperties(obj[propertyName], dataToSend));
      },
      error(e) {
        console.error(e);
      },
    });

    return obj;
  }
});
```

```js
// A mixin
import Ember from 'ember';
import queryDataUpdated from 'client/gql/subscriptions/dataUpdated';

export default Ember.Mixin.create({
  apollo: Ember.inject.service('apollo-service'),

  watchDataUpdated() {
    const subscriptionUpdated = this.get('apollo')
      .subscribe(queryDataUpdated, {}, 'data');

    this.set('subscriptionUpdated', subscriptionUpdated);
  },

  unwatchDataUpdated() {
    this.get('subscriptionUpdated').unsubscribe();
  },

  ondataUpdated: Ember.observer('subscriptionUpdated.data.timestamp', function () {
    const timestamp = this.get('subscriptionUpdated.data.timestamp');

    if (timestamp) {
      // Do something with your data
    }
  })
});
```

```graphql
// data-updated.graphql
subscription {
  dataUpdated {
    timestamp
  }
}
```

I don't know if it is the best way to do it, but it works well for my needs.

Aso, I had to use `subscriptions-transport-ws` version `0.8.3` because in the version 9.0+ the Subscription Manager is removed. On the server side you have to use a graphql version matching with `subscriptions-transport-ws`.

Another solution would be to update ember-apollo-client with last version of graphql, but it would need to recode some parts.

**Comment by xtagon** [comment]
Thanks to everyone here, and after trying things with newer versions of the libraries involved, I figured out how to make subscriptions work in an Ember + Apollo + Phoenix/Absinthe stack. Pretty cool! 😀 

Big thanks to @ArnaudValensi in particular for sharing your solution - only took some minor changes to work on my setup.

```javascript
import ApolloService from "ember-apollo-client/services/apollo";
import socketApolloLink from "npm:@absinthe/socket-apollo-link";
import AbsintheSocket from "npm:@absinthe/socket";
import Phoenix from "npm:phoenix";
import copyWithExtras from "ember-apollo-client/utils/copy-with-extras";
import EmberObject, { computed, setProperties } from "@ember/object";
import { run } from "@ember/runloop";
import config from "my-app/config/environment";

const phoenixSocket = new Phoenix.Socket(config.APP.phoenixSocketUrl);
const absintheSocket = AbsintheSocket.create(phoenixSocket);
const absintheSocketLink = socketApolloLink.createAbsintheSocketLink(absintheSocket);

export default ApolloService.extend({
  link: computed(function() {
    return absintheSocketLink;
  }),

  subscribe({query, variables}, propertyName = null) {
    const apolloClient = this.get("client");

    let observable = apolloClient.subscribe({query, variables});
    let subscription;

    let obj = EmberObject.create({
      unsubscribe() {
        subscription.unsubscribe();
      },
      [propertyName]: EmberObject.create({})
    });

    subscription = observable.subscribe({
      next: (rawData) => {
        if (rawData === undefined) {
          return;
        }

        let data = rawData[Object.keys(rawData)[0]];
        let dataToSend = copyWithExtras(data, [], []);

        run(() => setProperties(obj[propertyName], dataToSend));
      },
      error(e) {
        console.error(e);
      },
    });

    return obj;
  }
});
```

**Comment by bgentry** [comment]
Awesome!

Do you think there's a need to add a `subscribe` method to the query manager mixins? That could make it trivial to have a subscription that auto-cancels when the route/component/object that subscribed is torn down. It also seems like your `ApolloService.subscribe()` could go in the main Apollo service.

Anybody up for sending a PR?

**Comment by xtagon** [comment]
I would see that as useful, yes! In my experience there are really two use cases for subscribing though. In one case, you want to subscribe to an event (something happened) which has to be done with the apollo client directly. In the other case, you want to subscribe to an Ember object which just stays up to date with the latest results (which is what @ArnaudValensi shared). I think they should be two separate functions. In my mixin I ended up calling it `subscribeObject`.

While we're throwing ideas around, another thing that might be useful would be a pattern for doing a query immediately and then a subscription for subsequent results. That's mostly what I do when using subscriptions with the computed object pattern (not so much the event handling pattern).

Thoughts on what a decent PR might look like? I'm still pretty new to Apollo internals.

**Comment by coladarci** [comment]
I'm about to go down the copy-paste path to get this support - have there been any learnings worth sharing @xtagon ?

**Comment by xtagon** [comment]
@coladarci Not sure, other than my thoughts above. It's been a while since I last experimented with this. 

From what I remember, the code above worked pretty well. I'd love to try again with the latest versions of everything.

**Comment by coladarci** [comment]
Hey all - @xtagon 's code still works w/ latest versions. I have a pull that I'm finishing to this repo that wires up the subscribing / unsubscribing  in the `queryManager`. That said, I'm running into an annoying issue that started happening when switching to my fork (i.e when adding this code into the  `ember-apollo-client` project.

You see, on `init()`  a chain of `onInit`s trigger (`base-query-manager` -> `addon/apollo-service` -> `my-app/apollo-service`) all with the purpose of getting at  the `link` defined inside `my-app/apollo-service`. 

The trouble is that the user isn't authenticated `onInit`, so the phoenix socket can't connect (I'm passing a token). **I need to defer this link generation until the point where I can check the user's authentication status.**

I'm doing something similar for adding an auth header, using `setContext`, but I can't for the life of me figure out how to use that pattern to "defer" the link building until the user is authorized.  My understanding of `setContext` is that it provides properties to mutate the `context`, this isn't quite what I need to do...

This is what my link computed looks like (I'm already using phoenix sockets elsewhere, so I'm using the same socket, hence "channelClient"):
```
  link: computed(function () {
    const socket = this.get('channelClient.phoenixSocket');
    const absintheSocket = AbsintheSocket.create(socket);
    const absintheSocketLink = createAbsintheSocketLink(absintheSocket);

    const authLink = setContext((request, context) => this._runAuthorize(request, context));

    return authLink.concat(absintheSocketLink);
  }),
```

I'm very new to the GQL world, let alone this Apollo - if someone more seasoned as any ideas I think I am super close to finishing this pull! 

If there's another way to defer your project's `onInit` that would also get me there, but from my reading, links are supposed to work for this use-case.

**Comment by josemarluedke** [comment]
I believe we can closes this since #173 is now merged.
